### ISSUE TITLE
Zod resolved type change for Opaque type between v3.14.4 -> v3.15.1

### ISSUE THREAD

**Issue by bluepnume** [colinhacks/zod#1121]
I have the following code:

```typescript
type Opaque<Type, Token = unknown> = Type & { tag: Token }; 

enum FACE {
    QUEEN = 'QUEEN',
    KING = 'KING'
}

enum SUIT {
    HEARTS = 'HEARTS',
    DIAMONDS = 'DIAMONDS'
}

type FaceCard<Face extends FACE> = Opaque<`${ Face }_${ SUIT }`, 'FaceCard'>;

type Hand = z.output<
    z.ZodObject<{
        card : z.ZodType<FaceCard<FACE.QUEEN>, z.ZodTypeDef, string>,
    }>
>;
```

In `zod@3.14.4`, `Hand` resolved to:

```typescript
type Hand = {
    card: FaceCard<FACE.QUEEN>;
}
```

But now in `zoid@3.15.1`, `Hand` is resolving to:

```typescript
type Hand = {
    card: (FaceCard<FACE.QUEEN> | undefined) & FaceCard<FACE.QUEEN>;
}
```

Why did the second type get more complex here, when the first one is correct?

(By the way -- **incredible** library. Thank you for making this, seriously.)

**Comment by bentefay** [comment]
Hi @bluepnume. 

Sorry about this. I changed the definition of `z.ZodObject` in my PR (https://github.com/colinhacks/zod/issues/1121) that was recently merged. It's behaviour was unchanged in all the cases I tested, but it looks like you found a broken case. I've fixed the type to work with your example: 

https://github.com/colinhacks/zod/pull/1122

As a side node, although the type looks ugly and we should definitely fix the problem, I believe the types `{ property: (T | undefined) & T }` and `{ property: T }` are technically the same, so hopefully you shouldn't be seeing any issues with using the type.

**Comment by bluepnume** [comment]
Thanks @bentefay! Appreciate the fast turnaround!

No huge rush to merge, since I can just lock to `3.14.4` for now. But since you asked, the issue that this causes for me is demonstrated with the following code:

```typescript
type QueenCard = Hand['card'];

type CardValue = {
    [ FACE.QUEEN ] : 12,
    [ FACE.KING ] : 13,
};

type CardToValue<Face extends FACE> = Face extends keyof CardValue
    ? CardValue[Face]
    : never;

const cardToValue = <Face extends FACE>(card : FaceCard<Face>) : CardToValue<Face> => {
    return someLogic(card);
};

const value = cardToValue('' as QueenCard);

const expectedValue : 12 = value;

console.log('Value is', value, 'expected', expectedValue);
```

^ This code type-checks with `3.14.4`. But with `3.15.1`, I get:

```typescript
const expectedValue : 12 = value;
```

```
Type '12 | 13' is not assignable to type '12'.
```

-- essentially, while `(T | undefined) & T` and `T` are equivalent, you can use `T` as an index but you can't with `(T | undefined) & T`, so this type refinement breaks other parts of my code that rely on that.

**Comment by bentefay** [comment]
Ahh that makes complete sense. It's also not something I could have possibly dreamed up as a test :smile:. Now I see it though I suppose we could add that as a unit test, though it would be perhaps a little too specific? I'm not even sure what a minimal reproduction would be.

**Comment by colinhacks** [comment]
Resolved in 3.16, thanks @bentefay

**Comment by bluepnume** [comment]
Thanks @bentefay and @colinhacks! And again thanks for a stellar library, you guys rock.
